ledger-service/http-json/src/main/scala/com/digitalasset/http/PackageService.scala

```diff
@@ -370,11 +375,17 @@ object PackageService {
   final case class ContractTypeIdMap[CtId[T] <: ContractTypeId[T]](
       all: Map[RequiredPkg[CtId], ResolvedOf[CtId]],
       unique: Map[NoPkg[CtId], ResolvedOf[CtId]],
+      nameIds: Map[Ref.PackageName, String], // TODO(ETX-451): Multiple package ids per package name
   ) {
+    private def packageRefToPackageId(pkgRef: String): Option[String] = pkgRef match {
+      case s"#${name}" => nameIds.get(asPackageName(name))
+      case pkgId => Some(pkgId)
+    }
+
     private[http] def resolve(
         a: ContractTypeId[Option[String]]
     )(implicit makeKey: ContractTypeId.Like[CtId]): Option[ResolvedOf[CtId]] =
-      a.packageId match {
+      a.packageId.flatMap(packageRefToPackageId) match {
```

Comment:
It may be nice to introduce a typed package reference (like `com.daml.lf.data.TypeConRef`) that is used in place of `packageId` in most places (or until hitting the DB)

Reply:
Yes, I considered that, but didn't do it so far for a few reasons:
* `ContractTypeId`s are already a graph of different types, which differentiate over:
  * whether package ref part is supplied (definitely yes, definitely no, or maybe)
  * the kind of thing that's being pointed to (definitely a template, definitely an interface, either a template or interface and we know which one, unsure whether template or interface)
* While in principle this additional axis *could* be beneficial in distinguishing incompatible values, thus far I don't see a space where the distinction is actually relevant, beyond this one place, where it's already encoded in the `#` prefix.
* I'm conscious of not trying to make changes to this code-base that are overly sweeping. Given the horizon of this component, I tend to lean towards more targeted fixes and enhancements over larger refactorings.

Having said that, it might still be worth doing. That was just why I didn't do it in this PR.